File: code/scenes/src/main/java/org/fbme/scenes/controllers/diagram/DiagramController.java
package org.fbme.scenes.controllers.diagram;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Set;

public interface DiagramController<CompT, PortT, ConnT> {

    boolean isDiagramEditable();

    @NotNull
    PortController getPortController(@NotNull PortT port);

    @Nullable
    PortT findPort(int x, int y);

    @NotNull
    Set<CompT> getComponents();

    @NotNull
    Set<ConnT> getConnections();

    @NotNull
    Set<PortT> getPorts(@NotNull CompT component);

    @NotNull
    CompT getComponent(@NotNull PortT port);

    @NotNull
    PortT getSource(@NotNull ConnT edge);

    void setSource(@NotNull ConnT edge, @NotNull PortT port);

    @NotNull
    PortT getTarget(@NotNull ConnT edge);

    void setTarget(@NotNull ConnT edge, @NotNull PortT port);

    void removeEdge(@NotNull ConnT edge);

    @Nullable
    ConnT addEdge(@NotNull PortT sourcePort, @NotNull PortT targetPort);
}

File: code/scenes/src/main/kotlin/org/fbme/scenes/controllers/diagram/DiagramController.kt
package org.fbme.scenes.controllers.diagram

interface DiagramController<CompT, PortT, ConnT> {
    val isDiagramEditable: Boolean
    fun getPortController(port: PortT): PortController<PortT>
    fun getTemplateController(template: PortT): TemplateController<PortT>
    fun findPort(x: Int, y: Int): PortT?
    fun findPortTemplate(x: Int, y: Int): PortT?
    val components: Set<CompT>
    val connections: Set<ConnT>
    fun getPorts(component: CompT): Set<PortT>
    fun getComponent(port: PortT): CompT
    fun getSource(edge: ConnT): PortT?
    fun setSource(edge: ConnT, port: PortT)
    fun getTarget(edge: ConnT): PortT?
    fun setTarget(edge: ConnT, port: PortT)
    fun removeEdge(edge: ConnT)
    fun addEdge(sourcePort: PortT, targetPort: PortT): ConnT?
    fun addPort(port: PortT)
}
